model/src/util/image_hash.rs

```diff
@@ -0,0 +1,775 @@
+    ///
+    /// # Examples
+    ///
+    /// Refer to [`ImageHash::parse`]'s documentation for examples.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`ImageHashParseErrorType::Format`] error type if the
+    /// provided value isn't in a Discord image hash format. Refer to the
+    /// variant's documentation for more details.
+    ///
+    /// Returns an [`ImageHashParseErrorType::Range`] error type if one of the
+    /// hex values is outside of the accepted range. Refer to the variant's
+    /// documentation for more details.
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        Self::parse(value)
+    }
+}
+
+impl TryFrom<&str> for ImageHash {
+    type Error = ImageHashParseError;
+
+    /// Parse an image hash string into an efficient decimal store.
+    ///
+    /// # Examples
+    ///
+    /// Refer to [`ImageHash::parse`]'s documentation for examples.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`ImageHashParseErrorType::Format`] error type if the
+    /// provided value isn't in a Discord image hash format. Refer to the
+    /// variant's documentation for more details.
+    ///
+    /// Returns an [`ImageHashParseErrorType::Range`] error type if one of the
+    /// hex values is outside of the accepted range. Refer to the variant's
+    /// documentation for more details.
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Self::try_from(value.as_bytes())
+    }
+}
+
+/// Iterator over the [nibbles] of an image hash.
+///
+/// # Examples
+///
+/// Parse an image hash and then iterate over the nibbles:
+///
+/// ```
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// use twilight_model::util::ImageHash;
+///
+/// let input = b"1d9811c4cd3782148915c522b02878fc";
+/// let hash = ImageHash::parse(input)?;
+/// let mut nibbles = hash.nibbles();
+///
+/// assert_eq!(Some(b'1'), nibbles.next());
+/// assert_eq!(Some(b'd'), nibbles.next());
+/// assert_eq!(Some(b'c'), nibbles.nth(29));
+/// assert!(nibbles.next().is_none());
+/// # Ok(()) }
+/// ```
+///
+/// [nibbles]: https://en.wikipedia.org/wiki/Nibble
+#[derive(Debug)]
+pub struct Nibbles {
+    /// Current index of the iterator.
+    idx: usize,
+    /// Hash being iterated over.
+    inner: ImageHash,
+}
+
+impl Nibbles {
+    /// Create a new iterator over an image hash, creating nibbles.
+    const fn new(inner: ImageHash) -> Self {
+        Self {
+            idx: usize::MAX,
+            inner,
+        }
+    }
+
+    /// Advance the index in the iterator by the provided amount.
+    fn advance_idx_by(&mut self, by: usize) {
+        self.idx = if self.idx == usize::MAX {
+            0
+        } else {
+            let mut new_idx = self.idx.saturating_add(by);
+
+            if new_idx == usize::MAX {
+                new_idx -= 1;
+            }
+
+            new_idx
+        }
+    }
+
+    /// Parse the byte at the stored index.
+    ///
+    /// # Panics
```

Comment:
This section looks stale?

Reply:
is it?

Reply:
Maybe I’m confused. Im not clear where this panics given that it does its own bounds checking with `idx’.

Reply:
true. removed it and simplified the check